accepted/0003-interactive-audit-resolver.md

```diff
@@ -0,0 +1,155 @@
+# Audit resolutions
+
+## Summary
+
+Historically this proposal was titled "Interactive audit resolver". That idea was too wide to introduce in one go and parts of it are better served in userland packages. This is now the core subset of the original proposal containing support for resolutions file.
+
+This proposal is for adding means for a human to resolve issues from `npm audit` by making and documenting decisions to ignore particular false positives. 
+The implementation should introduce support for reading and applying decisions from `audit-resolve.json` file
```

Comment:
I would prefer either doing this through a more generic `audit.json` or through an `audit` property in package.json. Having a whole new file for a single purpose in the npm CLI feels… off.

There was discussion of additional config setups that I had with @darcyclarke that it could also potentially fit in, but they don’t exist yet so I can’t really recommend them.

Reply:
I didn't go with a package.json field because everyone kept saying we use it for too many things already and the audit file tends to be larger (from my practical experience of running the tool for 3-4 years)
Also, there's a plan to use it as vehicle for sharing recommendation. More on this here:
https://dev.to/naugtur/do-you-need-help-with-your-npm-audit-3olf

I'm not opinionated about the name. The final decision on the name will have to happen when we standardize the format in OpenJSF

Reply:
Other advantages of keeping it in a separate file include more granular codeowners targeting.

Reply:
I'll make it explicit here so it can be referred to in whatever OpenJSF work you're talking about: I'm a hard -1 on `audit-resolve.json`.

Reply:
audit.json is fine. 
OpenJSF work here 
https://github.com/openjs-foundation/pkg-vuln-collab-space

Reply:
@bnb on that file name, or on a separate file outside of package.json?

Reply:
@ljharb the specific file name.

Honestly, to me, the most preferable thing would be to support **both** `package.json` and an independent file to enable the widest possible set of end-users adopting this.

Reply:
I think having it in package.json, but ignored by package consumers, would be actively harmful (ie, be very surprising). package.json should be for things related to a package, which is usually published.

Reply:
@ljharb I'd say it'd be useful to check recursive packages for their own resolutions, but that's not exactly good security practice - users should actually be doing their due diligence here with their dependencies and that...doesn't exactly help them do that. (At best, it makes it far harder for them to audit dependencies accurately, and at worst users just blindly trust their dependencies' maintainers' decisions without even seeing what those decisions even were.)

So yeah, it shouldn't count as package metadata - it's more like *project* metadata than *package* metadata.

Reply:
https://dev.to/naugtur/do-you-need-help-with-your-npm-audit-3olf

Reply:
> that's not exactly good security practice

> users should actually be doing their due diligence here with their dependencies and that

I think it should be fully considered that a large number of projects will either:

- not have maintainers that have the bandwidth to do this
- not have maintainers that know how to do this
- not have maintainers that know this needs to be done at all
- not have maintainers (unmaintained / insufficiently maintained)

I work with a bunch of beginners every three months in our coding bootcamp at @upleveled and this is one example of maintainers who will not do this. And if you pick random packages on npm, more likely than not, you'll have something that is not well maintained.

So to save and check the audit decision as close as possible to the actual usage of the vulnerable code seems like actually a **better** security practice overall for the industry (at least as a fallback option).

Affecting sea change so that everyone in the industry spends time to check the security of their dependencies is unrealistic.

I think @naugtur's suggestion here could be one interesting way to implement this:

> package maintainers could publish ignore lists for the dependencies of their package when they did the research to verify the package is not affected.
>
> Such lists could be aggregated into recommendations maintained by security professionals.

To be clear, again, I'm not proposing this as the ONLY security measure - teams which have security processes, budget and practice in place should continue to do that and thus will be even more secure.

But for the majority of projects which will not have this level of focus on security, there should be a base level of security by default, provided by the ecosystem.

Reply:
The idea is to give people tools to do the best they can with the bandwidth they have. If we force them to address everything on their own, they're likely to just drop security from their pipeline

Reply:
And I'm just saying we should just not hide it by default, and force the user to take some explicit action, even if it's a compromise between security and time.

Reply:
We shouldn't - the user should never see something that isn't actionable for them.

Reply:
I'll agree to disagree and move on.